src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp

```diff
@@ -58,7 +58,87 @@ class NaiveConvolution
               const size_t dilationW = 1,
               const size_t dilationH = 1,
               const bool appending = false,
-              const typename std::enable_if_t<IsMatrix<InMatType>::value>* = 0)
+              const typename std::enable_if_t<IsMatrix<InMatType>::value &&
+                                              IsArma<InMatType>::value>* = 0)
+  {
+    using eT = typename InMatType::elem_type;
+    // Compute the output size.  The filterRows and filterCols computation must
+    // take into account the fact that dilation only adds rows or columns
+    // *between* filter elements.  So, e.g., a dilation of 2 on a kernel size of
+    // 3x3 means an effective kernel size of 5x5, *not* 6x6.
+    const size_t filterRows = filter.n_rows * dilationH - (dilationH - 1);
+    if (!appending)
+    {
+      const size_t filterCols = filter.n_cols * dilationW - (dilationW - 1);
+      const size_t outputRows = (input.n_rows - filterRows + dH) / dH;
+      const size_t outputCols = (input.n_cols - filterCols + dW) / dW;
+      output.zeros(outputRows, outputCols);
+    }
+
+    // Filter with dilation applied. Only dilated in the y direction.
+    FilMatType dilatedFilter;
+    if (dilationH == 1)
+    {
+      MakeAlias(dilatedFilter, filter, filter.n_rows, filter.n_cols);
+    }
+    else
+    {
+      dilatedFilter = FilMatType(filterRows, filter.n_cols,
+          GetFillType<FilMatType>::zeros);
+      for (size_t j = 0; j < filter.n_cols; j++)
+      {
+        eT* dilatedCol = dilatedFilter.colptr(j);
+        const eT* filCol = filter.colptr(j);
+        for (size_t i = 0; i < filter.n_rows; i++)
+        {
+          *dilatedCol = *(filCol++);
+          dilatedCol += dilationH;
+        }
+      }
+    }
+
+    for (size_t j = 0; j < output.n_cols; ++j)
+    {
+      for (size_t i = 0; i < output.n_rows; ++i)
+      {
+        eT accu = 0;
+        for (size_t kj = 0; kj < filter.n_cols; ++kj)
+        {
+          const eT* inCol = input.colptr(j * dW + kj * dilationW) + i * dH;
+          const eT* filCol = dilatedFilter.colptr(kj);
+          accu += arma::op_dot::direct_dot(filterRows, inCol, filCol);
```

Comment:
@nzachary The function `arma::op_dot::direct_dot()` is internal to Armadillo and shouldn't be used here.  The function can change or disappear entirely.

Extract from [Armadillo API policy](https://arma.sourceforge.net/docs.html#changelog):
> Any functionality within Armadillo which is not explicitly described in the public API documentation is considered as internal implementation detail, and may be changed or removed without notice. 

Only the functions and classes described in the [documentation](https://arma.sourceforge.net/docs.html) are considered as public API.